src/main/java/org/generationcp/middleware/util/VariableValueUtil.java

```diff
@@ -0,0 +1,109 @@
+
+			case NUMERIC_VARIABLE:
+				final StringBuilder range = new StringBuilder();
+				if (
+					(StringUtils.isNotBlank(variable.getMinValue()) //
+						|| StringUtils.isNotBlank(variable.getMaxValue()))) {
+					range.append(variable.getMinValue()).append("-");
+					range.append(variable.getMaxValue());
+
+				} else if (StringUtils.isNotBlank(variable.getScale().getMinValue()) //
+					|| StringUtils.isNotBlank(variable.getScale().getMaxValue())) {
+					range.append(variable.getScale().getMinValue()).append("-");
+					range.append(variable.getScale().getMaxValue());
+				}
+				return range.toString();
+
+			default:
+				return "";
+		}
+	}
+
+	//FIXME According with Mariano, observations should not accept invalid categories for a categorical scale
+	//FIXME invalidCategoricalScale should be removed when observations are fixed
+	private static boolean isValidValue(final Variable var, final String value, final boolean isMissingAccepted,
+		final boolean isOutOfBoundsCategoricalAccepted) {
+		if (StringUtils.isBlank(value)) {
+			return true;
+		}
+		if (var.getMinValue() != null && var.getMaxValue() != null) {
+			return validateIfValueIsMissingOrNumber(value.trim(), isMissingAccepted);
+		} else if (var.getScale().getDataType() == DataType.NUMERIC_VARIABLE) {
+			return validateIfValueIsMissingOrNumber(value.trim(), isMissingAccepted);
+		} else if (var.getScale().getDataType() == DataType.DATE_TIME_VARIABLE) {
+			return new DateValidator().isValid(value, "yyyyMMdd");
+		} else if (var.getScale().getDataType() == DataType.CATEGORICAL_VARIABLE) {
+			if (isOutOfBoundsCategoricalAccepted) {
+				return true;
+			}
+			return validateCategoricalValue(var, value);
+		}
+		return true;
+	}
+
+	private static boolean validateIfValueIsMissingOrNumber(final String value, final boolean isMissingAccepted) {
+		if (isMissingAccepted && MeasurementData.MISSING_VALUE.equals(value.trim())) {
+			return true;
+		}
+		return NumberUtils.isNumber(value);
+	}
+
+	private static boolean validateCategoricalValue(final Variable variable, final String value) {
+		final BigInteger categoricalValueId =
+			variable.getScale().getCategories().stream().filter(category -> value.equalsIgnoreCase(category.getName())).findFirst()
+				.map(category -> BigInteger.valueOf(category.getId())).orElse(null);
+		return categoricalValueId != null;
+	}
+
+	public static Integer resolveCategoricalValueId(final Variable variable, final String value) {
+		Integer categoricalValueId = null;
+		if (variable.getScale().getDataType().getId().equals(TermId.CATEGORICAL_VARIABLE.getId())) {
+			categoricalValueId =
+				variable.getScale().getCategories().stream().filter(category -> value.equalsIgnoreCase(category.getName())).findFirst()
+					.map(category -> BigInteger.valueOf(category.getId()).intValue()).orElse(null);
+			if (categoricalValueId == null) {
+				throw new MiddlewareRequestException("", "germplasm.attribute.invalid.categorical.value", variable.getName());
```

Comment:
Is this Util class meant to be for germplasm attribute only? If not, we should generalize the property code

Reply:
done